### ISSUE TITLE
[Phase 4] Loss Functions

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/speech-enhancement-system#5]
## Phase 4: Loss Functions

**Parent:** #1
**Depends on:** #3

### Objectives
Implement SI-SNR and spectral losses for training speech enhancement models.

### Tasks

- [ ] Implement SI-SNR loss function
- [ ] Implement spectral magnitude loss
- [ ] Implement combined/multi-resolution loss
- [ ] Verify losses are differentiable
- [ ] Test with known values
- [ ] Handle edge cases (silence, clipping)

### Files to Create/Modify

| File | Action | Description |
|------|--------|-------------|
| `src/losses.py` | Create | Loss function implementations |
| `tests/test_losses.py` | Create | Unit tests |

### Code Reference

```python
import torch
import torch.nn.functional as F

def si_snr_loss(estimate: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
    """Scale-Invariant SNR loss (negative, for minimization)."""
    # Zero-mean normalization
    estimate = estimate - estimate.mean(dim=-1, keepdim=True)
    target = target - target.mean(dim=-1, keepdim=True)

    # Projection
    dot = (estimate * target).sum(dim=-1, keepdim=True)
    s_target_energy = (target ** 2).sum(dim=-1, keepdim=True) + 1e-8
    s_target = dot * target / s_target_energy

    # Noise
    e_noise = estimate - s_target

    # SI-SNR
    si_snr = 10 * torch.log10(
        (s_target ** 2).sum(dim=-1) /
        ((e_noise ** 2).sum(dim=-1) + 1e-8) + 1e-8
    )

    return -si_snr.mean()


def spectral_loss(est_mag: torch.Tensor,
                  clean_mag: torch.Tensor,
                  alpha: float = 0.5) -> torch.Tensor:
    """Combined L1 and log-magnitude spectral loss."""
    l1 = F.l1_loss(est_mag, clean_mag)
    log_l2 = F.mse_loss(
        torch.log(est_mag + 1e-8),
        torch.log(clean_mag + 1e-8)
    )
    return alpha * l1 + (1 - alpha) * log_l2
```

### Definition of Done

- [ ] SI-SNR loss computes correct values (test with identical signals -> high SI-SNR)
- [ ] Losses are differentiable (backward() works)
- [ ] No NaN or Inf values in normal operation
- [ ] Edge cases handled (epsilon prevents division by zero)
- [ ] Unit tests pass

### Technical Notes

**For junior developers:**
- SI-SNR is scale-invariant: doesn't matter if signal is loud or quiet
- We return negative SI-SNR because PyTorch minimizes loss
- Always add small epsilon (1e-8) to prevent log(0) or division by zero
- Log-spectral loss emphasizes quieter frequency bins
